### PR TITLE
[BE] better type annotation for `torch.types`

### PR DESCRIPTION
**Pull request by XuehaiPan**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129559

Closes #129525

- #129525

cc @ezyang @malfet @xuzhao9 @gramster @huydhn

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -182,7 +182,10 @@ select = [
     "TCH004",
     "UP037", # ONNX does runtime type checking
 ]
-
+"torch/types.py" = [
+    "PYI042", # type alias should be CamelCase
```
Comment (Skylion007) [inline pyproject.toml:186]: Please use inline noqas for these instead.

File: torch/types.py [hunk 2]
```diff
@@ -14,100 +12,113 @@
...
+    DispatchKey as DispatchKey,
+    dtype as _dtype,
+    layout as _layout,
+    qscheme as _qscheme,
+    Size as Size,
+    SymBool as SymBool,
+    SymFloat as SymFloat,
+    SymInt as SymInt,
+    Tensor as Tensor,
+)
```
Comment (XuehaiPan) [inline torch/types.py:29]: `as` imports are better than type aliases defined via assignment.

File: torch/types.py [hunk 3]
```diff
@@ -14,100 +12,113 @@
...
+# bool or SymBool
+BoolLikeType: TypeAlias = Union[bool, SymBool]
 
 py_sym_types = (SymInt, SymFloat, SymBool)
-PySymType = Union[SymInt, SymFloat, SymBool]
+PySymType: TypeAlias = Union[SymInt, SymFloat, SymBool]
 
 # Meta-type for "numeric" things; matches our docs
-Number = Union[builtins.int, builtins.float, builtins.bool]
+Number: TypeAlias = Union[int, float, bool]
```
Comment (XuehaiPan) [inline torch/types.py:63]: Under the namespace of `torch.types`, the builtin classes (e.g. `int`) are not shadowed. This differs from the case under namespace `torch` where `builtins.int` is shadowed by `torch.int`. So here we do not need to use `builtins.int` or `_int`.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**